### PR TITLE
Fix #9504: Cache typeSize and covering set

### LINKED ISSUE #9504: Divergence checker does not detect loop with higher-kinded type, unlike Scala 2
## Minimized code

```Scala
trait Monad[F[_]] {
  def foo[A](fa: F[A]): Unit = {}
}

class Bla[F[_], A]

object Test {
  type Id[A] = A

  val bla: Bla[Id, Unit] = ???
  implicit def blaMonad[F[_]: Monad, S]: Monad[({type L[X] = Bla[F, X]})#L] = ???

  blaMonad.foo(bla)
}
```

## Output

With Dotty, the compiler runs for ~20 seconds before producing a giant error message:

```scala
-- Error: try/divloop.scala:19:10 ----------------------------------------------
19 |  blaMonad.foo(bla)
   |          ^
   |no implicit argument of type Monad[([_$3] =>> Any)] was found for an implicit parameter of method blaMonad in object Test.
   |I found:
   |
   |    Test.blaMonad[F, S](
   |      Test.blaMonad[F, S](
   |        Test.blaMonad[F, S](
...
```

## Expectation

With Scala 2, the compiler pretty much instantly stops with:
```scala
try/divloop.scala:19: error: diverging implicit expansion for type Monad[F]
starting with method blaMonad in object Test
  blaMonad.foo(bla)
  ^
```

This is not just a problem for malformed code: if I add another implicit to have a base case:
```scala
  implicit def idMonad: Monad[Id] = ???
```
Then Dotty still fails in the same way whereas Scala 2 is able to correctly resolve:
```scala
blaMonad(idMonad).foo(bla)
```
I came across this while trying to see what happens if cats Monad used dotty's extension method syntax (https://github.com/lampepfl/dotty/issues/9480).

Tentatively assigning to Miles since he wrote the divergence checker.

### PR DESCRIPTION
**Pull request by odersky**
The problem with #9504 is that types in the search history grow larger over time, since they contain type variables
that get instantiated with new types containing new type variables and so on. That's why divergence checking is fooled
into believing that the newly searched type is always smaller than the history.

To fix this, we need to take a snapshot of type size and covering set at the time the search history is
created.

This is much easier to do with the performance refactorings for search histories I did recently in #9409. As a first step, this commit contains those refactorings.

### CODE REVIEW

**Comment by odersky** [comment]
test performance please

**Comment by odersky** [comment]
test performance please

File: compiler/src/dotty/tools/dotc/typer/Implicits.scala [hunk 1]
```diff
@@ -239,23 +239,20 @@ object Implicits:
         Nil
       else
         val nestedCtx = ctx.fresh.addMode(Mode.TypevarsMissContext)
+        val candidates = new mutable.ListBuffer[Candidate]
+        var extensionOnly = true
```
Comment (smarter) [inline compiler/src/dotty/tools/dotc/typer/Implicits.scala:243]: I get that using a var here means that the `tryCandidate` closure can be reused, but the trade-off is that this var will become a BooleanRef and thus be allocated too, so I don't think we're gaining anything. Maybe instead `extensionOnly` and `tryCandidate` could be put in an object.

Comment (odersky) [reply compiler/src/dotty/tools/dotc/typer/Implicits.scala]: You are right. I think it's fine to use two closures instead.

**Pull request merged**